sympy/polys/series/tests/test_powerseriesring.py

```diff
@@ -0,0 +1,486 @@
+    """
+    domain = R.domain
+    domain_coeffs = []
+    for coeff in coeffs_l:
+        if isinstance(coeff, tuple):
+            domain_coeffs.append(domain(*coeff))
+        else:
+            domain_coeffs.append(domain(coeff))
+
+    return R.equal_repr(s, R.from_list(domain_coeffs, prec))
+
+
+@pytest.fixture(params=Ring_ZZ + Ring_QQ)
+def rd_int(request):
+    r = request.param
+    R = r()
+    x = R.gen
+    one = R.one
+
+    R3 = r(3)
+    x3 = R3.gen
+    one3 = R3.one
+
+    R10 = r(10)
+    x10 = R10.gen
+    one10 = R10.one
+    return R, x, one, R3, x3, one3, R10, x10, one10, r
+
+
+@pytest.fixture(params=Ring_QQ)
+def rd_rational(request):
+    r = request.param
+    R = r()
+    x = R.gen
+    one = R.one
+
+    R3 = r(3)
+    x3 = R3.gen
+    one3 = R3.one
+
+    R10 = r(10)
+    x10 = R10.gen
+    one10 = R10.one
+    return R, x, one, R3, x3, one3, R10, x10, one10, r
+
+
+def test_equal(rd_int):
+    R, *_ = rd_int
+    assert R.equal(R.from_list([1, 2, 3]), R.from_list([1, 2, 3])) is True
+    assert R.equal(R.from_list([1, 21, 3]), R.from_list([1, 2, 3])) is False
+    assert R.equal(R.from_list([1, 2, 3], 3), R.from_list([1, 2, 3], 3)) is None
+    assert R.equal(R.from_list([1, 2, 13], 3), R.from_list([1, 2, 3], 3)) is False
+    assert R.equal(R.from_list([1, 2, 3], 3), R.from_list([1, 2, 3], 10)) is None
+    assert R.equal(R.from_list([1, 21, 3], 3), R.from_list([1, 2, 3], 10)) is False
+    assert R.equal(R.from_list([1, 2, 3], 3), R.from_list([1, 2, 3, 4, 5], 10)) is None
+    assert R.equal(R.from_list([1, 2, 3, 4], None), R.from_list([1, 2, 3], 2)) is None
+    assert R.equal(R.from_list([1, 2, 3, 4], None), R.from_list([1, 1, 3], 2)) is False
+    assert R.equal(R.from_list([1, 2], None), R.from_list([1, 2, 3], 3)) is False
+
+
+def test_basics(rd_int):
+    R, *_, r = rd_int
+    R0 = r(0)
+    gen0 = R0.gen
+
+    assert R == r(6)
+    assert hash(R) == hash(r(6))
+    assert R.to_list(R.gen) == [0, 1], None
+    assert R0.equal_repr(R0.zero, R0.one)
+    assert R0.pretty(gen0) == "O(x**0)"
+    assert same(R0, gen0, [], 0)
+    assert same(R0, R0.multiply(gen0, gen0), [], 0)
+    assert same(R, R.add(R.from_list([2, 4, 5], 3), R.from_list([5], 2)), [7, 4], 2)
+
+
+def test_positive(rd_int):
+    R, *_ = rd_int
+    assert same(
+        R, R.positive(R.from_list([1, 2, 3, 4, 5, 6, 7], None)), [1, 2, 3, 4, 5, 6], 6
+    )
+
+
+def test_negative(rd_int):
+    R, *_ = rd_int
+    assert same(R, R.negative(R.gen), [0, -1], None)
+    assert same(
+        R,
+        R.negative(R.from_list([1, 2, 3, 4, 5, 6, 7], None)),
+        [-1, -2, -3, -4, -5, -6],
+        6,
+    )
+
+
+def test_int_add(rd_int):
+    _, _, _, R3, x3, one3, *_ = rd_int
+    assert same(R3, R3.add(x3, one3), [1, 1], None)
+    assert same(R3, R3.add(one3, R3.pow_int(x3, 4)), [1], 3)
+
+
+def test_rational_add(rd_rational):
+    R, _, one, *_ = rd_rational
+    assert same(R, R.add(R.negative(one), one), [], None)
```

Comment:
Why not combine tests like this? The test can be parametrised over all ring types both integer and rational.